pkg/common/cns-lib/vsphere/virtualcenter.go

```diff
@@ -228,11 +231,23 @@ func (vc *VirtualCenter) connect(ctx context.Context, requestNewSession bool) er
 		sessionMgr := session.NewManager(vc.Client.Client)
 		// SessionMgr.UserSession(ctx) retrieves and returns the SessionManager's CurrentSession field
 		// Nil is returned if the session is not authenticated or timed out.
-		if userSession, err := sessionMgr.UserSession(ctx); err != nil {
+		var userSession *types.UserSession
+		if userSession, err = sessionMgr.UserSession(ctx); err != nil {
 			log.Errorf("failed to obtain user session with err: %v", err)
 			return err
-		} else if userSession != nil {
-			return nil
+		}
+		if userSession != nil {
+			activeSession, err := sessionMgr.SessionIsActive(ctx)
+			if err != nil {
+				log.Errorf("failed to check user session is active. err: %v", err)
+				return err
+			}
+			if !activeSession {
+				return errors.New("session is not active")
+			}
+		} else {
+			// if session is NotAuthenticated, sessionMgr.UserSession(ctx) returns nil userSession with nil error
+			return errors.New("session is not authenticated")
```

Comment:
How does the system recover when this happens? Shouldn't we be creating a new session?

Reply:
yes, we are requesting a new session if this happens.

https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/503#discussion_r527211296